src/core/tests/unittests/serializer-polymorphic-test.cpp

```diff
@@ -0,0 +1,245 @@
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ *********************************************************************************/
+
+#include <warn/push>
+#include <warn/ignore/all>
+#include <gtest/gtest.h>
+#include <warn/pop>
+
+#include <inviwo/core/io/serialization/serialization.h>
+#include <inviwo/core/properties/ordinalproperty.h>
+#include <inviwo/core/properties/propertyfactory.h>
+#include <inviwo/core/common/inviwoapplication.h>
+
+#include <vector>
+#include <memory>
+
+namespace inviwo {
+
+TEST(SerializerPolymorphicTest, NullUniquePtr) {
+
+    std::unique_ptr<Property> prop1 = std::make_unique<FloatProperty>("float1", "float1", 0.123f);
+    prop1->setSerializationMode(PropertySerializationMode::All);
+
+    std::stringstream ss;
+    Serializer s("");
+
+    s.serialize("prop", prop1);
+    s.writeFile(ss);
+
+    std::unique_ptr<Property> prop2;
+
+    Deserializer d(ss, "");
+    d.registerFactory(InviwoApplication::getPtr()->getPropertyFactory());
+
+    d.deserialize("prop", prop2);
+
+    EXPECT_EQ(prop1->getClassIdentifier(), prop2->getClassIdentifier());
+    EXPECT_EQ(prop1->getIdentifier(), prop2->getIdentifier());
+
+    auto fprop1 = dynamic_cast<FloatProperty*>(prop1.get());
+    EXPECT_TRUE(fprop1);
+    auto fprop2 = dynamic_cast<FloatProperty*>(prop2.get());
+    EXPECT_TRUE(fprop2);
+
+    EXPECT_EQ(fprop1->get(), fprop2->get());
+}
+
+TEST(SerializerPolymorphicTest, NonNullUniquePtr) {
+    std::unique_ptr<Property> prop1 = std::make_unique<FloatProperty>("float1", "float1", 0.123f);
+    prop1->setSerializationMode(PropertySerializationMode::All);
+
+    std::stringstream ss;
+    Serializer s("");
+
+    s.serialize("prop", prop1);
+    s.writeFile(ss);
+
+    std::unique_ptr<Property> prop2 = std::make_unique<FloatProperty>("float2", "float2", 0.234f);
+    ;
+
+    Deserializer d(ss, "");
+    d.registerFactory(InviwoApplication::getPtr()->getPropertyFactory());
+
+    d.deserialize("prop", prop2);
+
+    EXPECT_EQ(prop1->getClassIdentifier(), prop2->getClassIdentifier());
+    EXPECT_EQ(prop1->getIdentifier(), prop2->getIdentifier());
+
+    auto fprop1 = dynamic_cast<FloatProperty*>(prop1.get());
+    EXPECT_TRUE(fprop1);
+    auto fprop2 = dynamic_cast<FloatProperty*>(prop2.get());
+    EXPECT_TRUE(fprop2);
+
+    EXPECT_EQ(fprop1->get(), fprop2->get());
+}
+
+TEST(SerializerPolymorphicTest, DifferentTypeUniquePtr) {
+    std::unique_ptr<Property> prop1 = std::make_unique<FloatProperty>("float1", "float1", 0.123f);
+    prop1->setSerializationMode(PropertySerializationMode::All);
+
+    std::stringstream ss;
+    Serializer s("");
+
+    s.serialize("prop", prop1);
+    s.writeFile(ss);
+
+    std::unique_ptr<Property> prop2 =
+        std::make_unique<DoubleProperty>("double2", "double2", 0.234f);
+    ;
+
+    Deserializer d(ss, "");
+    d.registerFactory(InviwoApplication::getPtr()->getPropertyFactory());
+
+    d.deserialize("prop", prop2);
```

Comment:
This makes sense, but what about DoubleProperty -> BoolProperty, which I think also would be valid given the setup here?

Reply:
Yes that should work fine. What would the problem be?

Reply:
It would be unexpected behaviour that it actually works?
It would mean that one has changed the type of a property, but not the identifier, and not provided a converter for the workspace?

Reply:
The use case for it is for example the camera, there we have a unique_ptr<Camera> and it can change between various child classes, Perspective, Orthographic etc. And we want to be able to switch between them.